rosbag2_py/src/rosbag2_py/_rosbag2_py.cpp

```diff
@@ -0,0 +1,177 @@
+          std::make_unique<rosbag2_compression::SequentialCompressionReader>();
+      reader_ =
+          std::make_unique<rosbag2_cpp::Reader>(std::move(sequential_reader));
+    } else {
+      throw std::runtime_error{"Reader class type " + reader_class +
+                               " not supported."};
+    }
+  }
+
+  void open(rosbag2_cpp::StorageOptions &storage_options,
+            rosbag2_cpp::ConverterOptions &converter_options) {
+    reader_->open(storage_options, converter_options);
+  }
+
+  bool has_next() { return reader_->has_next(); }
+
+  /// Return a tuple containing the topic name, the serialized ROS message, and
+  /// the timestamp.
+  pybind11::tuple read_next() {
+    const auto next = reader_->read_next();
+    rcutils_uint8_array_t rcutils_data = *next->serialized_data.get();
+    std::string serialized_data(
+        rcutils_data.buffer, rcutils_data.buffer + rcutils_data.buffer_length);
+    return pybind11::make_tuple(
+        next->topic_name, pybind11::bytes(serialized_data), next->time_stamp);
+  }
+
+  /// Return a mapping from topic name to topic type.
+  std::vector<rosbag2_storage::TopicMetadata> get_all_topics_and_types() {
+    return reader_->get_all_topics_and_types();
+  }
+
+private:
+  std::unique_ptr<rosbag2_cpp::Reader> reader_;
+};
+
+class Writer {
+public:
+  Writer(const std::string &writer_class) {
+    if (!writer_class.compare("SequentialWriter")) {
+      auto sequential_writer =
+          std::make_unique<rosbag2_cpp::writers::SequentialWriter>();
+      writer_ =
+          std::make_unique<rosbag2_cpp::Writer>(std::move(sequential_writer));
+    } else if (!writer_class.compare("SequentialCompressionWriter")) {
+      auto sequential_writer =
+          std::make_unique<rosbag2_compression::SequentialCompressionWriter>();
+      writer_ =
+          std::make_unique<rosbag2_cpp::Writer>(std::move(sequential_writer));
+    } else {
+      throw std::runtime_error{"Writer class type " + writer_class +
+                               " not supported."};
+    }
+  }
+
+  void open(rosbag2_cpp::StorageOptions &storage_options,
+            rosbag2_cpp::ConverterOptions &converter_options) {
+    writer_->open(storage_options, converter_options);
+  }
+
+  void create_topic(const rosbag2_storage::TopicMetadata &topic_with_type) {
+    writer_->create_topic(topic_with_type);
+  }
+
+  void remove_topic(const rosbag2_storage::TopicMetadata &topic_with_type) {
+    writer_->remove_topic(topic_with_type);
+  }
+
+  /// Write a serialized message to a bag file
+  void write(pybind11::tuple & message) {
+    auto bag_message = std::make_shared<rosbag2_storage::SerializedBagMessage>();
+
+    std::string data = std::string(pybind11::str(*message[1]));
+
+    bag_message->topic_name = std::string(pybind11::str(*message[0]));
+    bag_message->serialized_data = rosbag2_storage::make_serialized_message(&data, data.length());
+    bag_message->time_stamp = pybind11::cast<long>(*message[2]);
```

Comment:
Is it possible to cast to `rcutils_time_point_value_t`, which `time_stamp` is defined to be? In case the typedef of that custom type ever changes.

Reply:
fixed.